draft-vvv-webtransport-quic.md

```diff
@@ -154,62 +154,70 @@ datagram extension [QUIC-DATAGRAM].
 
-The client indication is a sequence of key-value pairs that are formatted in the
-following way:
+1. The client sends its headers on stream 0.
+1. The client sends FIN on stream 0.  This indicates that all headers have been
+   sent, and the server can now process them.
+1. The server processes the headers and makes access control decision based on
+   the origin.  Before this happens, the server MUST NOT process the
+   application data receiver from the client, or send its own header block.
+1. The server sends its header block on stream 0, followed by a FIN indicating
+   that the handshake has been successfully completed.
+
+The QuicTransport header stream payload SHALL be a header block.  A header
+block is a sequence of key-value pairs that are formatted in the following way:
 
 ~~~
- 0                   1                   2                   3
- 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|           Key (16)            |          Length (16)          |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                           Value (*)                         ...
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+QuicTransport Header {
+  Header Name Length (16),
+  Header Name (..),
+  Header Value Length (16),
+  Header Value (..),
+}
 ~~~
-{: #fig-frame title="Client indication format" :}
+{: #fig-frame title="QuicTransport Header Format" :}
 
 The pair includes the following fields:
 
-Key:
-
-: Indicates the field that is being expressed.
-
-Length:
+Name:
 
-: Indicates the length of the value (the length of the key and the length
-  itself are not included).
+: Indicates the name of the field that is being expressed.  The name of the
+  field is case-insensitive, and when encoded on the wire, it MUST consist of
+  ASCII characters in range from `a` to `z`, `0` to `9`, `-`, or `.`.  It MUST
+  begin with an alphanumeric character.
 
 Value:
 
-: The value of the field, the semantics of which are determined by the key.
+: The value of the header field, the semantics of which are determined by the
+  name.
 
-A FIN on the stream 2 SHALL indicate that the message is complete.  The client
-MUST send the entirety of the client indication and a FIN immediately after
-opening the connection.  The server MUST NOT process any application data before
-receiving the entirety of the client indication.  The total length of the client
-indication MUST NOT exceed 65,535 bytes.
+A FIN on the stream 0 SHALL indicate that the message is complete.  The client
+MUST send the entirety of the header block and a FIN immediately after opening
+the connection.  The server MUST NOT process any application data before
+receiving the entirety of the headers.  The total length of a header block MUST
+NOT exceed 1,048,576 bytes; implementations MAY impose additional limit on the
```

Comment:
Additional implementation-specific limits can lead to interop issues. Could we instead say that servers MUST accept headers up to that length?

Reply:
We could, but in practice implementations will inevitable impose their own limit, and we can't preemptively enforce this, so I don't think it makes sense.

Reply:
This seems just like the setting for MAX_HEADER_LIST size. Unfortunately that setting has legacy baggage from decades of HTTP deployment. It seems with a new application protocol, the server could defe a limit thats enforceable. That helps cut down the memory commitment and is interoperable. Exchanging such a setting might require something ALPS-like though.

Reply:
I do think that we should resolve those potential interop issues - and we can enforce this by using padding right?

Reply:
I don't think we want to send one megabyte worth of padding.  I've filed #20 to track this.